In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import pyfpgrowth

pd.set_option('display.max_columns', 100)
warnings.filterwarnings('ignore')
sns.set()

In [2]:
def parse_file(filename):
    # open file and read the content of the file
    with open(filename,'r') as data:
        content = data.read()
        # get unique items in the dataset
        items = content.replace('\n',',').split(',')
        items.remove('')  #remove blank rows/transaction
        items = set(items)
        
        # get all line by line transactions
        all_lines = content.split('\n')
        all_lines.remove('')  #remove blank rows/transaction
        no_of_line = len(all_lines)
        print('There are {} unique items in the dataset'.format(len(items)))
        
        # create a dataframe where if Item is present in Transaction in dataset then its value is 1
        raw_data = pd.DataFrame(data=0,index=range(0,no_of_line),columns=items)
        for index,line in enumerate(all_lines):
            print("Parsing {}/{} transaction".format(index,no_of_line),end='\r')
            item_list = line.strip().split(',')
            raw_data.loc[index,item_list] = 1   
    return raw_data

In [3]:
raw_data = parse_file('groceries.csv')
raw_data.sample(3)

There are 169 unique items in the dataset


packaged fruit/vegetables  meat spreads  soft cheese  beef  \
9209                          0             0            0     0   
8675                          0             0            0     0   
2385                          0             0            0     0   

      potted plants  prosecco  flower (seeds)  sauces  baby cosmetics  \
9209              0         0               0       0               0   
8675              0         0               0       0               0   
2385              0         0               0       0               0   

      brown bread  liquor  semi-finished bread  cereals  dish cleaner  \
9209            0       0                    0        0             0   
8675            0       0                    0        0             0   
2385            0       0                    0        0             0   

      organic sausage  baby food  pork  cookware  yogurt  sausage  \
9209                0          0     0         0       0        0   
8675                0          0     1         0       0        0   
2385                0          0     0         0       0        0   

      make up remover  misc. beverages  male cosmetics  grapes  \
9209                0                0               0       0   
8675                0                0               0       0   
2385                0                0               0       0   

      frozen potato products  long life bakery product  bottled beer  \
9209                       0                         1             0   
8675                       0                         0             0   
2385                       0                         0             0   

      sparkling wine  pickled vegetables  rum  mayonnaise  cocoa drinks  fish  \
9209               0                   0    0           0             0     0   
8675               0                   0    0           0             0     0   
2385               0                   0    0           0             0     0   

      canned vegetables  chocolate marshmallow  soap  oil  brandy  margarine  \
9209                  0                      0     0    0       0          0   
8675                  0                      0     0    0       0          0   
2385                  0                      0     0    0       0          0   

      red/blush wine  coffee  turkey  specialty fat  cling film/bags  napkins  \
9209               0       0       0              0                0        0   
8675               0       1       0              0                0        1   
2385               0       1       0              0                0        0   

      salt  flour  canned beer  specialty cheese  specialty chocolate  ...  \
9209     0      0            0                 0                    0  ...   
8675     0      0            0                 0                    0  ...   
2385     0      0            0                 0                    0  ...   

      roll products  cake bar  domestic eggs  honey  snack products  \
9209              0         0              0      0               0   
8675              0         0              0      0               0   
2385              0         0              0      0               0   

      light bulbs  whisky  bathroom cleaner  condensed milk  photo/film  tea  \
9209            0       0                 0               0           0    0   
8675            0       0                 0               0           0    0   
2385            0       0                 0               0           0    0   

      chicken  curd cheese  pasta  baking powder  bottled water  berries  \
9209        1            0      0              0              0        0   
8675        0            0      0              0              0        0   
2385        0            0      0              0              0        0   

      kitchen utensil  cream  hygiene articles  candles  dessert  \
9209                0      0                 0       

In [4]:
raw_data.shape

(9835, 169)

In [5]:
# create copy ofr raw_data
raw_data_copy = raw_data.copy()

### Top 10 products

In [6]:
# Top 10 products
raw_data.sum().nlargest(10)

whole milk          2513
other vegetables    1903
rolls/buns          1809
soda                1715
yogurt              1372
bottled water       1087
root vegetables     1072
tropical fruit      1032
shopping bags        969
sausage              924
dtype: int64

### Lowest 10 Products

In [7]:
# Lowest 10 Products
raw_data.sum().nsmallest(10)

baby food                1
sound storage medium     1
preservation products    2
bags                     4
kitchen utensil          4
baby cosmetics           6
frozen chicken           6
toilet cleaner           7
make up remover          8
whisky                   8
dtype: int64

### finding items which makes 80% of the sales

In [8]:
# finding items which makes 80% of the sales
total_purchase = raw_data.sum().sum()
item_sale_cumulative = raw_data.sum().div(total_purchase).mul(100).sort_values(ascending=False).cumsum()
item_sale_cumulative.head(3)

whole milk           5.794729
other vegetables    10.182858
rolls/buns          14.354232
dtype: float64

In [9]:
sales_eighty_per = item_sale_cumulative[item_sale_cumulative <= 80]
sales_eighty_per.tail(5)

butter milk      77.383725
specialty bar    78.004012
beverages        78.594323
ham              79.184633
meat             79.770332
dtype: float64

In [10]:
print("There are {} items which account for 80% of the sales".format(sales_eighty_per.shape[0]))
unique_item_per = round(sales_eighty_per.shape[0]/len(raw_data.columns),4)*100
print("Only {}% out of item all the items  account for 80% of the sales".format(unique_item_per))

There are 53 items which account for 80% of the sales
Only 31.36% out of item all the items  account for 80% of the sales


### To reduce the size of the dataset we will only consider the transactions with these 53 items which account for 80% of sales

In [11]:
# consider only 53 items for further analysis
subset = raw_data[sales_eighty_per.index].copy()

# remove rows all zeros value
subset = subset.loc[subset.sum(axis=1)!=0,:].copy()

subset.sample(3)

whole milk  other vegetables  rolls/buns  soda  yogurt  bottled water  \
9161           0                 0           1     1       1              0   
4276           1                 0           1     0       0              0   
7654           0                 0           0     0       0              0   

      root vegetables  tropical fruit  shopping bags  sausage  pastry  \
9161                0               0              0        0       0   
4276                0               0              0        0       0   
7654                0               0              0        0       0   

      citrus fruit  bottled beer  newspapers  canned beer  pip fruit  \
9161             0             0           1            0          0   
4276             0             0           0            0          0   
7654             0             0           0            1          0   

      fruit/vegetable juice  whipped/sour cream  brown bread  domestic eggs  \
9161                      0                   0            0              0   
4276                      0                   0            0              0   
7654                      0                   0            0              0   

      frankfurter  margarine  coffee  pork  butter  curd  beef  napkins  \
9161            0          0       0     0       1     0     0        0   
4276            0          0       0     0       0     0     0        0   
7654            0          0       0     1       0     0     0        0   

      chocolate  frozen vegetables  chicken  white bread  cream cheese  \
9161          0                  0        0            0             0   
4276          0                  0        0            0             0   
7654          0                  0        0            0             0   

      waffles  salty snack  long life bakery product  dessert  sugar  \
9161        1            0                         0        0      0   
4276        0            0                         0        0      0   
7654        0            0                         0        0      0   

      UHT-milk  berries  hamburger meat  hygiene articles  onions  \
9161         0        0               0                 0       0   
4276         0        0               0                 0       0   
7654         0        0               0                 0       0   

      specialty chocolate  candy  misc. beverages  frozen meals  oil  \
9161                    0      0                0             0    0   
4276                    0      0                0             0    0   
7654                    0      0                0             0    0   

      butter milk  specialty bar  beverages  ham  meat  
9161            0              0          0    0     0  
4276            0              0          0    0     0  
7654            0              0          0    0     0

In [12]:
print("Orignal Transaction dataset shape {}".format(raw_data.shape))
print("Reduce Transaction dataset shape {}".format(subset.shape))

Orignal Transaction dataset shape (9835, 169)
Reduce Transaction dataset shape (9372, 53)


### Creating Cross Sell Grid Matrix where row is Primary Product and columns is other products purhcase with primary

In [14]:
# creating empty cross sell matrix 
cross_sell_grid = pd.DataFrame(index=subset.columns,columns=subset.columns,data=0)

# add primary count for the count of reference product considered/index
cross_sell_grid.insert(loc = 0, column='Primary_Count', value=0)

# filling the cross sell matrix
for item in subset.columns:
    sales = subset[subset[item]==1]
    
    # adding count for reference product
    cross_sell_grid.at[item,'Primary_Count'] = sales[item].sum()
    
    # adding count for all other products i.e. secondary product
    count = sales.sum()
    cross_sell_grid.loc[item,count.index] = count
    
    # making column of primary product 0 as it is already counted in Primary_Count column
    cross_sell_grid.at[item,item] = 0

In [15]:
cross_sell_grid

Primary_Count  whole milk  other vegetables  \
whole milk                         2513           0               736   
other vegetables                   1903         736                 0   
rolls/buns                         1809         557               419   
soda                               1715         394               322   
yogurt                             1372         551               427   
bottled water                      1087         338               244   
root vegetables                    1072         481               466   
tropical fruit                     1032         416               353   
shopping bags                       969         241               228   
sausage                             924         294               265   
pastry                              875         327               222   
citrus fruit                        814         300               284   
bottled beer                        792         201               159   
newspapers                          785         269               190   
canned beer                         764          87                89   
pip fruit                           744         296               257   
fruit/vegetable juice               711         262               207   
whipped/sour cream                  705         317               284   
brown bread                         638         248               184   
domestic eggs                       624         295               219   
frankfurter                         580         202               162   
margarine                           576         238               194   
coffee                              571         184               132   
pork                                567         218               213   
butter                              545         271               197   
curd                                524         257               169   
beef                                516         209               194   
napkins                             515         194               142   
chocolate                           488         164               125   
frozen vegetables                   473         201               175   
chicken                             422         173               176   
white bread                         414         168               135   
cream cheese                        390         162               135   
waffles                             378         125                99   
salty snack                         372         110               106   
long life bakery product            368         133               105   
dessert                             365         135               114   
sugar                               333         148               106   
UHT-milk                            329          39                80   
berries                             327         116               101   
hamburger meat                      327         145               136   
hygiene articles                    324         126                94   
onions                              305         119               140   
specialty chocolate                 299          79                60   
candy                               294          81                68   
misc. beverages                     279          69                55   
frozen meals                        279          97                74   
oil                                 276         111                98   
butter milk                         275         114               102   
specialty bar                       269          64                55   
beverages                           256          67                51   
ham                                 256         113                90   
meat                                254          98                98   

                          rolls/buns  soda  yogurt  bottled water  \
whole milk    

# Naive cross sell recommendations basis cross sell product grid count

In [18]:
# function for getting cross sell recommendations basis cross grid count
def cross_sell_recommend(cross_sell_grid, primary_item):
    primary = cross_sell_grid.loc[primary_item,:].copy()
    primary.drop('Primary_Count',inplace=True)
    xsell_recommend = primary[primary > 0].nlargest(3).copy()
    return xsell_recommend

In [19]:
# get cross sell recommendations for soda
xsell_recommend = cross_sell_recommend(cross_sell_grid,'soda')
xsell_recommend

whole milk          394
rolls/buns          377
other vegetables    322
Name: soda, dtype: int64

In [24]:
# get cross sell recommendations for other candy
xsell_recommend = cross_sell_recommend(cross_sell_grid,'candy')
xsell_recommend

soda          85
whole milk    81
rolls/buns    70
Name: candy, dtype: int64

### Association Rule Mining using FP Growth

In [25]:
# prepraring data for FP Growth : Getting Items purchased sequence for each transactions
subset.columns[subset.loc[0,:] > 0]

Index(['citrus fruit', 'margarine'], dtype='object')

In [26]:
transactions = []
for idx,row in subset.iterrows():
    item_purchase = subset.columns[row > 0].tolist()
    transactions.append(item_purchase)  

In [27]:
# Setting mimium support threshold to be 2% of the transactions
support_threshold_per = 0.5
support_threshold_count = int(support_threshold_per*subset.shape[0]/100)
support_threshold_count

46

In [28]:
patterns = pyfpgrowth.find_frequent_patterns(transactions, support_threshold_count)
frequent_pattern = pd.DataFrame.from_dict(patterns,
                                          orient='index',
                                          columns=['Support_Count']).sort_values(by='Support_Count',ascending=False)
frequent_pattern.head(10)

Support_Count
(whole milk,)                                 2513
(other vegetables,)                           1903
(other vegetables, whole milk)                 736
(rolls/buns, whole milk)                       557
(whole milk, yogurt)                           551
(root vegetables, whole milk)                  481
(other vegetables, root vegetables)            466
(other vegetables, yogurt)                     427
(other vegetables, rolls/buns)                 419
(tropical fruit, whole milk)                   416

From Frequent Pattern we can see that (other vegetables, whole milk) are the most frequent product brought together

In [29]:
frequent_pattern.shape

(869, 1)

In [30]:
confidence_threshold = 0.6
rules = pyfpgrowth.generate_association_rules(patterns, confidence_threshold)
association_rule = pd.DataFrame.from_dict(rules,orient='index',
                                          columns=['Secondary_Product',
                                                   'Confidence']).sort_values(by='Confidence',ascending=False)
association_rule

Secondary_Product  \
(curd, domestic eggs)                                     (whole milk,)   
(butter, curd)                                            (whole milk,)   
(other vegetables, pip fruit, root vegetables)            (whole milk,)   
(rolls/buns, root vegetables, yogurt)                     (whole milk,)   
(citrus fruit, root vegetables, whole milk)         (other vegetables,)   
(citrus fruit, other vegetables, yogurt)                  (whole milk,)   
(other vegetables, pip fruit, yogurt)                     (whole milk,)   
(domestic eggs, pip fruit)                                (whole milk,)   
(curd, pip fruit)                                         (whole milk,)   
(domestic eggs, margarine)                                (whole milk,)   
(butter, domestic eggs)                                   (whole milk,)   
(other vegetables, tropical fruit, yogurt)                (whole milk,)   
(fruit/vegetable juice, other vegetables, yogurt)         (whole milk,)   
(pip fruit, root vegetables, whole milk)            (other vegetables,)   
(other vegetables, root vegetables, whipped/sou...        (whole milk,)   
(other vegetables, root vegetables, yogurt)               (whole milk,)   
(bottled water, butter)                                   (whole milk,)   

                                                    Confidence  
(curd, domestic eggs)                                 0.734375  
(butter, curd)                                        0.716418  
(other vegetables, pip fruit, root vegetables)        0.675000  
(rolls/buns, root vegetables, yogurt)                 0.647887  
(citrus fruit, root vegetables, whole milk)           0.633333  
(citrus fruit, other vegetables, yogurt)              0.626667  
(other vegetables, pip fruit, yogurt)                 0.625000  
(domestic eggs, pip fruit)                            0.623529  
(curd, pip fruit)                                     0.623377  
(domestic eggs, margarine)                            0.621951  
(butter, domestic eggs)                               0.621053  
(other vegetables, tropical fruit, yogurt)            0.619835  
(fruit/vegetable juice, other vegetables, yogurt)     0.617284  
(pip fruit, root vegetables, whole milk)              0.613636  
(other vegetables, root vegetables, whipped/sou...    0.607143  
(other vegetables, root vegetables, yogurt)           0.606299  
(bottled water, butter)                               0.602273

There is around 73% chance that whole milk will be bought with curd and  domestic eggs. Hence it there can three strategy that can be used here:
1. Put Whole Milk near Curd and Domestic Eggs 
2. Bundle Whole Milk, Curd, Domestic Eggs as a single product offering
3. Put Whole Milk at the opposite end of Curd and Domestic Eggs so that Customer will be force to travel to the other end and in the process may some additional product